### ISSUE TITLE
Gloo Edge 1.13.x with IstioSDS 1.14 doesn't mount spiffe volumes

### ISSUE THREAD

**Issue by edubonifs** [solo-io/gloo#8092]
### Gloo Edge Version

1.13.x (beta)

### Kubernetes Version

None

### Describe the bug

Hi, 

We have a customer that wants to install Gloo Edge 1.13.8 with Istio integration, Istio version is 1.14.6. The problem is that they don't want all the pods from Gloo-system to have a sidecar injected, they only want the gateway-proxy pod to have a sidecar injected.

If you label Gloo-system namespace and install GE with istioSDS enabled everything works fine, but for installing GE with the sidecar injected on Gateway Proxy pod only, we found the following problem:

### Steps to reproduce the bug

If you install with GE with the following command:
```
helm install gloo glooe/gloo-ee --namespace gloo-system \
  --create-namespace --set-string license_key=LICENSE_KEY --version v1.13.8 --values values.yaml
```
And this values.yaml file:

```
global:
  istioIntegration:
    whitelistDiscovery: true
    enableIstioSidecarOnGateway: true
  istioSDS:
    enabled: true
    customSidecars:
      - name: istio-proxy
        image: 'docker.io/istio/proxyv2:1.14.6'
```

The gateway fails getting the certificates mounted:

```
{"level":"fatal","ts":"2023-04-14T13:18:38.721Z","logger":"sds_server","caller":"cmd/main.go:84","msg":"could not find file '/etc/istio-certs/key.pem'","version":"1.13.7","stacktrace":"main.main\n\t/workspace/gloo/projects/sds/cmd/main.go:84\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250"}
```

The only possible way I have seen of injecting a sidecar proxy only to the gateway proxy without labeling the full namespace is using this values.yaml file, in which clients would need to specify spiffe volumes to be added:

```
global:
  istioIntegration:
    labelInstallNamespace: true
    whitelistDiscovery: true
    enableIstioSidecarOnGateway: true
  istioSDS:
    enabled: true
    customSidecars:
      - name: istio-proxy
        image: 'docker.io/istio/proxyv2:1.14.6'
        args:
          - proxy
          - sidecar
          - '--domain'
          - $(POD_NAMESPACE).svc.cluster.local
          - '--configPath'
          - /etc/istio/proxy
          - '--binaryPath'
          - /usr/local/bin/envoy
          - '--serviceCluster'
          - istio-proxy-prometheus
          - '--drainDuration'
          - 45s
          - '--parentShutdownDuration'
          - 1m0s
          - '--discoveryAddress'
          - 'istiod.istio-system.svc:15012'
          - '--proxyLogLevel=warning'
          - '--proxyComponentLogLevel=misc:error'
          - '--connectTimeout'
          - 10s
          - '--proxyAdminPort'
          - '15000'
          - '--controlPlaneAuthPolicy'
          - NONE
          - '--dnsRefreshRate'
          - 300s
          - '--statusPort'
          - '15021'
          - '--trust-domain=cluster.local'
          - '--controlPlaneBootstrap=false'
        env:
          - name: OUTPUT_CERTS
            value: /etc/istio-certs
          - name: JWT_POLICY
            value: third-party-jwt
          - name: PILOT_CERT_PROVIDER
            value: istiod
          - name: CA_ADDR
            value: 'istiod.istio-system.svc:15012'
          - name: ISTIO_META_MESH_ID
            value: cluster.local
          - name: ISTIO_META_CLUSTER_ID
            value: Kubernetes
          - name: POD_NAME
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.name
          - name: POD_NAMESPACE
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
          - name: INSTANCE_IP
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: status.podIP
          - name: SERVICE_ACCOUNT
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: spec.serviceAccountName
          - name: HOST_IP
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: status.hostIP
          - name: CANONICAL_SERVICE
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: 'metadata.labels[''service.istio.io/canonical-name'']'
          - name: CANONICAL_REVISION
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: 'metadata.labels[''service.istio.io/canonical-revision'']'
          - name: PROXY_CONFIG
            value: |
              {"meshId":"cluster.local"}
          - name: ISTIO_META_POD_NAME
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.name
          - name: ISTIO_META_CONFIG_NAMESPACE
            valueFrom:
              fieldRef:
                apiVersion: v1
                fieldPath: metadata.namespace
          - name: TRUST_DOMAIN
            value: cluster.local
          - name: ISTIO_META_NETWORK
            value: kube-network
        resources:
          limits:
            cpu: '2'
            memory: 1Gi
          requests:
            cpu: 100m
            memory: 128Mi
        volumeMounts:
          - name: istiod-ca-cert
            mountPath: /var/run/secrets/istio
          - name: istio-envoy
            mountPath: /etc/istio/proxy
          - name: istio-certs
            mountPath: /etc/istio-certs/
          - name: istio-token
            mountPath: /var/run/secrets/tokens
          - name: workload-socket
            mountPath: /var/run/secrets/workload-spiffe-uds
          - name: workload-certs
            mountPath: /var/run/secrets/workload-spiffe-credentials
        readinessProbe:
          httpGet:
            path: /healthz/ready
            port: 15021
            scheme: HTTP
          initialDelaySeconds: 1
          timeoutSeconds: 1
          periodSeconds: 2
          successThreshold: 1
          failureThreshold: 30
        imagePullPolicy: IfNotPresent
gloo:
  gatewayProxies:
    gatewayProxy:
      extraVolumes:
      - name: workload-socket
      - name: workload-certs
```

### Expected Behavior

I think we would need a way of injecting just gateway-proxy deployment without having the label the full namespace

**Comment by SantoDE** [comment]
Linked to #8621

**Comment by bewebi** [comment]
I’m starting to look into this issue and I wanted to clarify what the problem/ask is

It seems that at some point, newer versions of Istio started to require additional volumes which we don’t have in our templates
(I haven’t been able to find documentation for these other than [here](https://istio.io/latest/docs/ops/integrations/spire/) where one of the required paths is referenced, so if there is some clearer documentation that states exactly what volumes are expected, that would be great)

It also seems there is a workaround, which is to use the `customSidecar` section to inject config for the new volumes (along with everything else that we usually include by default)

I want to confirm that it’s not that there is some functionality that’s broken, but rather that specifying a custom sidecar is cumbersome, because all of the config that you get for free when not using a custom sidecar needs to be included in the values file? And also, perhaps, that these are now standard requirements for Istio and therefore should be included?
Is there anything more to it?

It seems we could probably add those volumes without it breaking any older versions of Istio which don’t require them, so I’m hoping that’s quite simple

**Comment by bewebi** [comment]
Upon further review, the issue here seems to be that users aren't specifying an up-to-date Istio version for the sidecar.

The Helm value for this is in a confusing location, as the image specification for the `istio-proxy` container for the Gateway deployment is taken from `global.glooMtls.istioProxy.image`
This currently defaults to v1.9.5 which is no longer supported
When a supported version is specified (without specifying a custom sidecar) the gateway pod consistently comes up healthily

I think this is the preferred solution compared to specifying a custom sidecar
If a custom sidecar is specified, all of the config for the istio-proxy container must be included in-line, including everything that typically comes for free when not using a custom sidecar

Give this, I'm inclined to close #8666, which has helm changes which seem to be ineffectual, and open a slimmer PR which bumps the default Istio version to one we support and which exposes the `image` blocks for the `istioSDS` setting under the  `istioSDS` block, doing so in such a way as to not break existing value usage

Will proceed upon sign-off from field

**Comment by kondasoulas** [comment]
Let me also add our experience on this matter in case you find it helpful.
 
We currently have a k8s cluster (version `1.24`) and we are using Istio `1.18.1` and gloo edge `1.14.12`.

So far our settings in values.yaml was as following:

```
gloo:
  global:
    istioIntegration:
      enableIstioSidecarOnGateway: true
    istioSDS:
      enabled: true
    glooMtls:
      enabled: true
      istioProxy:
        image:
          registry: docker.io/istio
          repository: proxyv2
          tag: 1.18.1
          pullPolicy: IfNotPresent
```

With the above configuration our `gateway-proxy` pods are being injected with an `istio-proxy` sidecar of the version we want (the same as the service mesh) and they are getting to **running** state without issues. We also have to note that in this way an _init container_ named `istio-init` is also present in the `gateway-proxy` pods.

During our testing we realised that the above configuration **breaks** our implementation for **canary** deployments (with `subsets` in the upstreams definition). To further elaborate although we target specifically (with a header) only the canary pods we receive answers from the non-canary ones. Our `fallbackPolicy` in our upstreams is set to `ANY_ENDPOINT` and if we go the extra step and update it to `NO_FALLBACK` instead of getting answers from our non-canary pods we get "no healthy upstream". 

As a result we get the impression that using the above setup in values.yaml the canary pods (using subsets) **cannot be reached** at all. However, all the other functionality of our Gloo Edge gateway stays intact and we have no issues with Istio `1.18.1`. This problematic behaviour is also reproduced with Gloo Edge `1.15` as well as Istio `1.17` (and all the related possible Gloo Edge/Istio combinations). In all cases the subset functionality is broken.

After some testing we realised that the subset functionality breaks when the `istio-init` init container is present in the `gateway-proxy` pods. So we tried to find a way to have the `istio-proxy` sidecar present in `gateway-proxy` pods but not the `istio-init` init container. The only way we managed to do that is by updating our values.yaml file as follows:

```
gloo:
  global: 
    istioSDS:
      enabled: true
      customSidecars:
        - name: istio-proxy
          image: docker.io/istio/proxyv2:1.18.1
          args:
            - proxy
            - sidecar
            - '--domain'
            - $(POD_NAMESPACE).svc.cluster.local
          [remaining istio-proxy declaration here]
    glooMtls:
      enabled: true
```

of course also with the additions for the SPIFFE volumes:

```
  gatewayProxies:
    gatewayProxy:
      extraVolumes:
        - name: workload-socket
        - name: credential-socket
```
In this way `istio-init` init container is not present in our `gateway-proxy` pods, `istio-proxy` sidecar is in the version we want it to be, and more importantly the subsets functionality is back to working as it should. 

@bewebi we would like to bring this into your attention since you are working on a related PR (https://github.com/solo-io/gloo/pull/8696) that could help us avoid this complexity. Can you think of any possible reason for the `istio-init` init container to impact the subsets functionality in this way? We can of course share the `istio-init` definition if necessary to verify its actions.

**Comment by bewebi** [comment]
A fix for this is encompassed in the work done for #8621 
This has been released in OSS v1.15.10 and will will be included in OSS 1.15.10, EE 1.15.5, and EE 1.16.0-beta1

In addition, backports of the fix to all LTS branches have merged or will soon and will be included in the next patch release for each minor release

**Comment by bewebi** [comment]
This fix has been released in the latest LTS branches for both OSS and EE
Specifically OSS 1.12.59, 1.13.29, 1.14.23, 1.15.10, and 1.16.0-beta15 and EE 1.12.59, 1.13.27, 1.14.11, 1.15.5, and 1.16.0-beta1
